notebook/static/notebook/js/notebook.js

```diff
@@ -2846,7 +2846,93 @@ define([
             this._checkpoint_after_save = false;
         }
     };
-    
+
+    Notebook.prototype.save_notebook_as = function() {
+        var that = this;
+        var current_dir = $('body').attr('data-notebook-path').split('/').slice(0, -1).join("/");
+        current_dir = current_dir? current_dir + "/": "";
+        var dialog_body = $('<div/>').append(
+            $('<p/>').addClass('save-message')
+                .text(i18n.msg._('Enter a notebook path relative to notebook dir'))
+        ).append(
+            $('<br/>')
+        ).append(
+            $('<input/>').attr('type','text').attr('size','25')
+            .attr('data-testid', 'save-as')
+            .addClass('form-control')
+        );
+
+        var d = dialog.modal({
+            title: 'Save As',
+            body: dialog_body,
+            keyboard_manager: this.keyboard_manager,
+            notebook: this,
+            buttons: {
+                Cancel: {},
+                Save: {
+                    class: 'btn-primary',
+                    click: function() {
+                        var nb_path = d.find('input').val();
+                        var nb_name = nb_path.split('/').slice(-1).pop();
+                        // If notebook name does not contain extension '.ipynb' add it
+                        var ext = utils.splitext(nb_name)[1];
+                        if (ext === '') {
+                            nb_name = nb_name + '.ipynb';
+                            nb_path = nb_path + '.ipynb';
+                        }
+                        var save_thunk = function() {
+                            var model = {
+                                'type': 'notebook',
+                                'content': that.toJSON(),
+                                'name': nb_name
+                             };
+                            return that.contents.save(nb_path, model)
+                                .then(function(data) {
+                                    d.modal('hide');
+                                    that.notebook_name = data.name;
+                                    that.notebook_path = data.path;
+                                    that.session.rename_notebook(data.path);
+                                    that.events.trigger('notebook_renamed.Notebook', data);
+
+                                },function(error) {
+                                   const msg = i18n.msg._(error.message || 'Unknown error saving notebook');
+                                   $('.save-message').html(`<span style='color:red;'>${msg}</span>`);
+                                });
+                        }
+                        that.contents.get(nb_path, {type: 'notebook'}).then(function(data) {
```

Comment:
This should at least use `content: false` to avoid getting the full content of the other notebook if it does exist. It's also not great in that it assumes any error getting a model means that the file is not there.

The right way to do this is probably to add a parameter when saving, to ask the server to throw an error if the filename already exists. Maybe this way of doing it is good enough for now, though. I'd like to get some other opinions on it.

Reply:
> This should at least use content: false to avoid getting the full content of the other notebook if it does exist.  

This is done, thank you.

> It's also not great in that it assumes any error getting a model means that the file is not there.
I will try to think of a better way.  Maybe we go back to something like my original implementation which used a flag to toggle between copy/save on the server..